extensions/amp-date-countdown/1.0/date-countdown.js

```diff
@@ -0,0 +1,247 @@
+import {useResourcesNotify} from '../../../src/preact/utils';
+
+const NAME = 'DateCountdown';
+
+// Constants
+/** @const {number} */
+const MILLISECONDS_IN_DAY = 24 * 60 * 60 * 1000;
+
+/** @const {number} */
+const MILLISECONDS_IN_HOUR = 60 * 60 * 1000;
+
+/** @const {number} */
+const MILLISECONDS_IN_MINUTE = 60 * 1000;
+
+/** @const {number} */
+const MILLISECONDS_IN_SECOND = 1000;
+
+/** @const {number} */
+const DELAY = 1000;
+
+/** @const {Object<string, number>} */
+const TimeUnit = {
+  DAYS: 1,
+  HOURS: 2,
+  MINUTES: 3,
+  SECONDS: 4,
+};
+
+// Default prop values
+const DEFAULT_OFFSET_SECONDS = 0;
+const DEFAULT_LOCALE = 'en';
+const DEFAULT_WHEN_ENDED = 'stop';
+const DEFAULT_BIGGEST_UNIT = 'DAYS';
+
+/**
+ * @param {!DateCountdownPropsDef} props
+ * @return {PreactDef.Renderable}
+ */
+export function DateCountdown({
+  endDate,
+  timeleftMs,
+  timestampMs,
+  timestampSeconds,
+  offsetSeconds = DEFAULT_OFFSET_SECONDS,
+  whenEnded = DEFAULT_WHEN_ENDED,
+  locale = DEFAULT_LOCALE,
+  biggestUnit = DEFAULT_BIGGEST_UNIT,
+  render,
+  children,
+  ...rest
+}) {
+  useResourcesNotify();
+  const {'playable': playable} = useAmpContext();
+  const epoch = useMemo(
+    () =>
+      new Date(
+        getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) +
+          offsetSeconds * DELAY
+      ),
+    [endDate, timeleftMs, timestampMs, timestampSeconds, offsetSeconds]
+  );
+  const [timeLeft, setTimeLeft] = useState(epoch - Date.now());
+  const localeStrings = useMemo(() => getLocaleWord(locale), [locale]);
+  const rootRef = useRef(null);
+
+  useEffect(() => {
+    if (!playable) {
+      return;
+    }
+    const win = rootRef.current.ownerDocument.defaultView;
+    const interval = win.setInterval(() => {
+      const newTimeLeft = epoch - Date.now() + DELAY;
+      setTimeLeft(newTimeLeft);
+      if (whenEnded === DEFAULT_WHEN_ENDED && newTimeLeft < 1000) {
+        win.clearInterval(interval);
+      }
+    }, DELAY);
+    return () => win.clearInterval(interval);
+  }, [playable, epoch, whenEnded]);
+
+  const data = {
+    ...getYDHMSFromMs(timeLeft, biggestUnit),
+    ...localeStrings,
+  };
+  return (
+    <div ref={rootRef} {...rest}>
+      {render(data, children)}
+    </div>
+  );
+}
+
+/**
+ * Calculate the epoch time that this component should countdown to from
+ * one of multiple input options.
+ * @param {string|undefined} endDate
+ * @param {number|undefined} timeleftMs
+ * @param {number|undefined} timestampMs
+ * @param {number|undefined} timestampSeconds
+ * @return {number}
+ */
+function getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) {
+  let epoch;
+
+  if (endDate) {
+    epoch = Date.parse(endDate);
+  } else if (timeleftMs) {
+    epoch = Date.now() + timeleftMs;
+  } else if (timestampMs) {
+    epoch = timestampMs;
+  } else if (timestampSeconds) {
+    epoch = timestampSeconds * 1000;
+  }
+
+  if (epoch === undefined) {
```

Comment:
What is the expected behavior when multiple are provided? e.g. both `endDate` and `timeleftMs`